### ISSUE TITLE
Moving from forces that inherit from types to forces with type attributes

### ISSUE THREAD

**Issue by ctk3b** [shirtsgroup/InterMol#240]
Currently intermol `Force` classes are subclasses of `ForceType` objects some of which are subclasses of `AbstractForceType` objects. The idea was
* `AbstractForceType` objects contain information applicable to a set of forces
  * E.g., `AbstractBondType` knows `bondingtype1`, `bondingtype2`, `order` and `c`
* `ForceType` knows _parameter_ values for a specific functional form
  * E.g., `HarmonicBondType` knows `length` and `k`
* `Force` knows which atoms are involved
  * E.g., `HarmonicBond` knows `atom1` and `atom2`

For GROMACS and LAMMPS this means:
* Read type directives and create `ForceTypes`
* Read forces, lookup types, extract parameters from types, create forces as subclass of that type

The other way to handle this is to have `Force` classes with `ForceType` attributes. Preliminary work for making this switch is on the `forcetype` branch. When switching to this style, there are simply `Bond`, `Angle` etc. `Forces` that know `atom1-n` as opposed to specific subclasses for every type. 

For GROMACS and LAMMPS this means:
* Read type directives and create `ForceTypes`
* Read forces, create the more general forces, lookup type and set it as the attribute

---

From messing around in the `forcetype` branch, the second approach simplifies the GROMACS and LAMMPS readers a bit because there's less looking up and passing around of parameters. It also simplifies the canonical function slightly since they only need to deal with `ForceTypes` and don't need to discern between those and `Forces` depending on what's passed to them.  It would also simplify things like swapping types of a force or perhaps to have a force with more than one type?

I personally also find this model simpler to digest mentally.

@mrshirts, if I remember correctly, you had some concerns with how this approach would impact the Desmond parser. Could you outline those below?

**Comment by mrshirts** [comment]
I'm still trying to wrap my mind around this, since I have been thinking about it the other way :)

So, in the new version, ForceTypes are created, but forces are not subclasses of them -- instead, the ForceType is an attribute of the Force?

**Comment by ctk3b** [comment]
Exactly. For example:  the force `Bond` with attributes `atom1` and `atom2` has another attribute `HarmonicBondType` which could be the same as before.

**Comment by swails** [comment]
This is very similar to the ParmEd model.  The basic model I use there is that a `Structure` contains a list of atoms, residues, bonds, angles, dihedrals, ... and other valence connectivities.  And each of those valence connectivities contains attributes `atom1`, `atom2`, ..., `atomN` which are references to the relevant objects in the `Structure.atoms` list.  And each valence connectivity object has a `type` attribute that is the relevant parameter type (e.g., `BondType`, `AngleType`, ...) that contains the relevant force field parameters (or `None` if they're not available).

If this is a feasible model to transition to, it would probably make integration with ParmEd that much easier.  I also think there are legitimate use-cases for separating a valence term from its parameters conceptually.

**Comment by mrshirts** [comment]
Looking over the Desmond code I don't necessarily see anything wrong with this.  Desmond only uses types for pair interactions; bonds, angles and dihedrals only exist as forces.

We should use this opportunity to sychronize across codes the process of taking a text string and turning it into an interaction -- there are a couple of variants across the codes as well.
